### PR TITLE
New zipfilesystem module and FixedLengthSource class

### PR DESCRIPTION
**Pull request by swankjesse**
The FixedLengthSource will be necessary with the ZIP implementation. When
reading an uncompressed slice we need to put a bound on the end of the
source. When reading a compressed slice we expect it to exhaust exactly
as expected.

### CODE REVIEW

File: okio-zipfilesystem/src/jvmTest/kotlin/okio/zipfilesystem/FixedLengthSourceTest.kt [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+ */
+package okio.zipfilesystem
+
+import okio.Buffer
+import okio.IOException
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.Test
+import kotlin.test.fail
+
+internal class FixedLengthSourceTest {
```
Comment (bnorm) [inline okio-zipfilesystem/src/jvmTest/kotlin/okio/zipfilesystem/FixedLengthSourceTest.kt:24]: Thinking about a test case along the lines of:
1. Source of length 16, wrap with non-truncated fixed size of 10
2. Read 10 (`bytesReceived` now equal to fixed size)
3. Read 10 and expect error and no data

All the existing tests look to put `bytesReceived` over the size and never equal. I don't think there is a bug or anything, but it might be nice to have a test for this boundary condition.

File: okio-zipfilesystem/src/jvmMain/kotlin/okio/zipfilesystem/FixedLengthSource.kt [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+/**
+ * A source that returns [size] bytes of [delegate].
+ *
+ * This throws an [IOException] if the delegate returns fewer than [size] bytes.
+ *
+ * If [truncate] is true, this truncates to [size] bytes. Otherwise this requires that [delegate]
+ * will return exactly [size] bytes, and will throw an [IOException] if it doesn't.
+ */
+internal class FixedLengthSource(
```
Comment (JakeWharton) [inline okio-zipfilesystem/src/jvmMain/kotlin/okio/zipfilesystem/FixedLengthSource.kt:31]: Can this ship in the core behind an extension on `Source` so that it can be used in OkHttp and places like length-delimited proto streams with Wire?

Comment (swankjesse) [reply okio-zipfilesystem/src/jvmMain/kotlin/okio/zipfilesystem/FixedLengthSource.kt:31]: Yes!

File: okio-zipfilesystem/src/jvmMain/kotlin/okio/zipfilesystem/FixedLengthSource.kt [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+    if (result != -1L) bytesReceived += result
+
+    // Throw an exception if we received too few bytes or too many.
+    if ((bytesReceived < size && result == -1L) || bytesReceived > size) {
+      if (result > 0L && bytesReceived > size) {
+        // If we received bytes beyond the limit, don't return them to the caller.
+        sink.truncateToSize(sink.size - (bytesReceived - size))
+      }
+      throw IOException("expected $size bytes but got $bytesReceived")
```
[the remark this answers is not in the document]
Comment (swankjesse) [reply okio-zipfilesystem/src/jvmMain/kotlin/okio/zipfilesystem/FixedLengthSource.kt:65]: I don’t think so? The numbers will tell the story

File: okio-zipfilesystem/src/jvmTest/kotlin/okio/zipfilesystem/FixedLengthSourceTest.kt (hunk 1 above)
Comment (swankjesse) [reply okio-zipfilesystem/src/jvmTest/kotlin/okio/zipfilesystem/FixedLengthSourceTest.kt:24]: Good call! Added this fencepost test.

File: okio-zipfilesystem/build.gradle [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+      dependencies {
+        implementation deps.test.junit
+        implementation deps.test.assertj
+        implementation deps.kotlin.test.jdk
+      }
+    }
+  }
+}
+
+// modify these lines for MANIFEST.MF properties or for specific bnd instructions
```
Comment (oldergod) [inline okio-zipfilesystem/build.gradle:25]: Sentence

**Review by oldergod — APPROVED** [review]

File: okio-zipfilesystem/build.gradle (hunk 4 above)
Comment (swankjesse) [reply okio-zipfilesystem/build.gradle:25]: Can do. Follow-up!

File: okio-zipfilesystem/src/jvmMain/kotlin/okio/zipfilesystem/FixedLengthSource.kt [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+  override fun read(sink: Buffer, byteCount: Long): Long {
+    // Figure out how many bytes to attempt to read.
+    //
+    // If we're truncating, we never attempt to read more than what's remaining.
+    //
+    // Otherwise we expect the underlying source to be exactly the promised size. Read as much as
+    // possible and throw an exception if too many bytes are returned.
+    val toRead = when {
+      bytesReceived > size -> 0L // Already read more than the promised size.
```
Comment (yschimke) [inline okio-zipfilesystem/src/jvmMain/kotlin/okio/zipfilesystem/FixedLengthSource.kt:46]: Why isn't this an error? Doesn't this state turn a presumed while loop in a caller into an infinite loop? e.g. If there are multiple concurrent threads calling this class?

File: okio-zipfilesystem/build.gradle (hunk 4 above)
Comment (swankjesse) [reply okio-zipfilesystem/build.gradle:25]: The follow-up got rid of this comment! https://github.com/square/okio/pull/921/files#
